openstack/standalone-cinder/pkg/provisioner/provisioner.go

```diff
@@ -74,19 +77,37 @@ func NewCinderProvisioner(client kubernetes.Interface, id, configFilePath string
 // Provision creates a storage asset and returns a PV object representing it.
 func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
 	var (
-		connection volumeservice.VolumeConnection
-		mapper     volumeMapper
-		pv         *v1.PersistentVolume
-		cleanupErr error
+		connection        volumeservice.VolumeConnection
+		mapper            volumeMapper
+		pv                *v1.PersistentVolume
+		volumeID          string
+		err               error
+		cleanupErr        error
+		useExistingVolume bool
+		volumeState       string
 	)
 
 	if options.PVC.Spec.Selector != nil {
 		return nil, fmt.Errorf("claim Selector is not supported")
 	}
 
+	if volumeID, useExistingVolume = options.PVC.Annotations[ShouldProvisionVolumeAnn]; useExistingVolume {
+		volumeState, err = p.vsb.getCinderVolumeState(p.VolumeService, volumeID)
+		if err != nil {
+			glog.Errorf("Failed to get volume %s, err: %v", volumeID, err)
+			goto ERROR
+		}
+		if volumeState != "available" {
+			glog.Errorf("Volume %s state is not available, current state %v", volumeID, volumeState)
+			err = fmt.Errorf("Volume %s state is not available, current state %v", volumeID, volumeState)
+			goto ERROR
+		}
+		goto CONNECT
+	}
+
 	// TODO: Check access mode
 
-	volumeID, err := p.vsb.createCinderVolume(p.VolumeService, options)
+	volumeID, err = p.vsb.createCinderVolume(p.VolumeService, options)
```

Comment:
refactor `createCinderVolume` to pass volumeID to `source_volid`  in CreateOpts.

Reply:
i am also interested to see how cinder create volume works if you pass both source volume id and capacity. Cinder API [spec](https://developer.openstack.org/api-ref/block-storage/v3/index.html#volumes-volumes) says `The API creates a new volume with the same size as the source volume.`, but doesn't specify the behavior if both capacity and source volume id are set.

Reply:
@rootfs I've tried with cinder cli, must set size in createOpts, if both size and source volume id are set, cinder will still set according to size.

Reply:
This is not to clone an existing volume, but use the existing volume and create pv with the volume information. No actual volume provisioning here.

Reply:
The only caveat to that is the newly requested size must be equal to or greater than the source volume size.

So if just source volume is provided, Cinder will create the new volume to match the source size. If both are specified and the requested size is at least the same size as the source volume, that new volume will be created from the source volume but sized as requested.